**Notas para contenedor de docker:**

Comando de docker para ejecución de la nota de forma local:

nota: cambiar `<ruta a mi directorio>` por la ruta de directorio que se desea mapear a `/datos` dentro del contenedor de docker.

```
docker run --rm -v <ruta a mi directorio>:/datos --name jupyterlab_numerical -p 8888:8888 -d palmoreck/jupyterlab_numerical:1.1.0
```

password para jupyterlab: `qwerty`

Detener el contenedor de docker:

```
docker stop jupyterlab_numerical
```


Documentación de la imagen de docker `palmoreck/jupyterlab_numerical:1.1.0` en [liga](https://github.com/palmoreck/dockerfiles/tree/master/jupyterlab/numerical).

---

Nota generada a partir de [liga1](https://drive.google.com/file/d/1zCIHNAxe5Shc36Qo0XjehHgwrafKSJ_t/view), [liga2](https://drive.google.com/file/d/12L7rOCgW7NEKl_xJbIGZz05XXVrOaPBz/view).

In [61]:
!pip3 install --user -q cvxpy

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
You should consider upgrading via the 'pip install --upgrade pip' command.


# Problemas de optimización convexa con restricciones lineales de igualdad

En esta nota se consideran resolver problemas de optimización con restricciones lineales de igualdad de la forma:

$$\min f_o(x)$$

$$\text{sujeto a:} Ax=b$$

con variable de optimización $x \in \mathbb{R}^{n}$ y $A \in \mathbb{R}^{p \times n}, b \in \mathbb{R}^p$ conocidos. 

Se asume lo siguiente:

* $f:\mathbb{R}^n \rightarrow \mathbb{R}$ convexa y $\mathcal{C}^2(\text{dom}f_o)$.

* $rank(A) = p < n$: tenemos menos restricciones que variables y los renglones de $A$ son linealmente independientes. 

* Existe un punto óptimo $x^*$ por lo que el problema tiene solución y el valor óptimo se denota por $p^* = f_o(x^*) = \inf f_o(x)$.

* Los puntos iniciales $x^{(0)}$ de los métodos iterativos están en $\text{dom}f_o$ y los conjuntos $f_o(x^{(0)})$-subnivel son conjuntos cerrados.

Ver [1.4.Polinomios_de_Taylor_y_diferenciacion_numerica](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/I.computo_cientifico/1.4.Polinomios_de_Taylor_y_diferenciacion_numerica.ipynb) y [4.1.Optimizacion_numerica_y_machine_learning](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.1.Optimizacion_numerica_y_machine_learning.ipynb) para definiciones utilizadas en esta nota.

## Método de Newton aplicado a las [condiciones de Karush-Kuhn-Tucker](https://en.wikipedia.org/wiki/Karush%E2%80%93Kuhn%E2%80%93Tucker_conditions) (KKT) de optimalidad

El problema del inicio lo nombramos **primal** (**se recomienda leer el ápendice de esta nota donde se dan definiciones utilizadas en esta sección.)**

En el caso de optimización sin restricciones **la condición** que nos ayudó a determinar si **un punto $x^*$ es óptimo** fue $\nabla f(x^*) = 0$ (ver definiciones al final de [4.1.Optimizacion_numerica_y_machine_learning](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.1.Optimizacion_numerica_y_machine_learning.ipynb) y la nota [4.2.Algoritmos_para_UCO](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.2.Algoritmos_para_UCO.ipynb)) . Tal condición es **necesaria y suficiente** por la suposición que $f_o$ es convexa diferenciable. Esta condición es en **general** un conjunto de $n$ **ecuaciones no lineales** en $n$ variables que resuelve el problema sin restricciones.

Para los problemas de **optimización convexos** con **restricciones lineales o afín de igualdad** las **condiciones de Karush-Kuhn-Tucker de optimalidad** constituyen un conjunto de condiciones **necesarias y suficientes** que ayudan a resolver al problema primal. Tales condiciones KKT de optimalidad para este problema son:

$$Ax^* = b$$

$$\nabla f(x^*) + A^T\nu^*=0$$

**Comentarios:** 

* La **ecuación de factibilidad primal** ($1$a ecuación anterior) es una ecuación **lineal**.

* La **ecuación de factibilidad dual** ($2$a ecuación anterior) en general es una ecuación **no lineal**.

* Por las condiciones KKT de optimalidad: $x^* \in \text{dom}f_o$ es óptimo **si y sólo si** $Ax^*=b$ y  $\exists \nu^* \in \mathbb{R}^p$ tal que $\nabla f(x^*) + A^T\nu^*=0$. Entonces resolver el sistema de ecuaciones anteriores para las variables $x^*$ y $\nu ^*$ equivale a resolver el problema de optimización primal.

* Los métodos para resolver el problema primal consisten en:

    * Eliminar las restricciones de igualdad a partir de transformaciones al problema primal y utilizar [4.2.Algoritmos_para_UCO](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.2.Algoritmos_para_UCO.ipynb).
    * Plantear y resolver el **problema dual** y utilizar [4.2.Algoritmos_para_UCO](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.2.Algoritmos_para_UCO.ipynb).
    * **Extender** el método de Newton para problemas sin restricciones (ver [4.2.Metodo_de_Newton_Python](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.2.Metodo_de_Newton_Python.ipynb)) y resolver directamente las condiciones KKT de optimalidad.

En esta nota utilizamos el enfoque número $3$: extender el método de Newton. Para una revisión de los otros dos enfoques ver [liga2](https://drive.google.com/file/d/12L7rOCgW7NEKl_xJbIGZz05XXVrOaPBz/view)

## Método de Newton con punto inicial factible

El método de Newton sin restricciones revisado en [4.2.Metodo_de_Newton_Python](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.2.Metodo_de_Newton_Python.ipynb) se extiende  para el caso de restricciones de igualdad considerando:

* El punto inicial $x^{(0)}$ **debe ser factible:** $x^{(0)} \in \text{dom}f_o$ y $Ax^{(0)}=b$ (aunque esta restricción puede eliminarse con el método de Newton para **puntos no factibles**, ver [4.4.Metodo_de_Newton_para_puntos_iniciales_no_factibles_Python](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.4.Metodo_de_Newton_para_puntos_iniciales_no_factibles_Python.ipynb)).

* El paso de Newton $\Delta x_{\text{nt}}$ **debe modificarse** de modo que satisfaga la ecuación de factibilidad primal.



El método de Newton con punto inicial factible puede obtenerse de varias formas. Ejemplos son:

* Aproximación de segundo orden vía el teorema de Taylor.

* Linearización de las condiciones KKT de optimalidad también vía el teorema de Taylor.

Para una introducción al teorema de Taylor ver [1.4.Polinomios_de_Taylor_y_diferenciacion_numerica](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/I.computo_cientifico/1.4.Polinomios_de_Taylor_y_diferenciacion_numerica.ipynb).

### Aproximación de segundo orden

Supóngase que $x$ es punto factible del problema primal. Por las suposiciones del inicio se cumple por el teorema de Taylor:

$$\hat{f}(x+v) = f(x) + \nabla f(x)^Tv + \frac{1}{2}v^T \nabla ^2 f(x) v$$

es una aproximación de **segundo orden** a $f(x)$ con centro en $x$ y variable $v$.

El método de Newton para puntos factibles consiste en considerar resolver el problema:



$$\displaystyle \min_{v \in \mathbb{R}^n} \hat{f}(x+v)$$ 

$$\text{sujeto a: } A(x+v) = b$$

para la variable de optimización $v \in \mathbb{R}^n$. Tal problema es un problema de optimización convexa con **función objetivo cuadrática**.

Diferentes condiciones determinan si el problema anterior tiene solución y si es única. Tales condiciones tienen una **relación directa** con las condiciones KKT de optimalidad que son:

$A(x + v^*) = b$  y por tanto $Av^*=0$ (factibilidad primal) y $\nabla ^2f(x) v^* + \nabla f(x) + A^T w^* = 0$ para $w^*$ variable óptima dual asociada al problema.


Reescribiendo las condiciones anteriores en un sistema nombrado **sistema KKT para el problema primal** con **matriz KKT** se tiene:


$$\begin{array}{l}
\left[ \begin{array}{cc}
\nabla^2f(x) & A^T\\
A & 0
\end{array}
\right] \cdot 
\left[\begin{array}{c}
v\\
w
\end{array}
\right] =
\left[\begin{array}{c}
-\nabla f(x)\\
0
\end{array}
\right]
\end{array}
$$

**Comentarios:**

* El **paso de Newton**, $\Delta x _{\text{nt}}$ se define como la **solución única** del sistema anterior con matriz KKT **no singular**. En este caso $v^* = \Delta x_{\text{nt}}$ y la solución **analítica** es:


$$\begin{array}{l}
\left[\begin{array}{c}
\Delta x_{\text{nt}}\\
w^*
\end{array}
\right] =
\left[ \begin{array}{cc}
\nabla^2f(x) & A^T\\
A & 0
\end{array}
\right] ^{-1}
\left[\begin{array}{c}
-\nabla f(x)\\
0
\end{array}
\right]
\end{array}
$$

* Si la función objetivo $f_o$ del problema primal es cuadrática, entonces la actualización $x + \Delta x_{\text{nt}}$ resuelve de forma **exacta** tal problema y la variable $w$ es la variable óptima dual. Si la función objetivo $f_o$ es **cercana** a ser cuadrática, $x + \Delta x_{\text{nt}}$ y $w$ deben ser **muy buenas estimaciones** del óptimo primal $x^*$ y del óptimo dual $\nu ^*$ respectivamente.

# Constrained Equality Convex Optimization (CECO)

Se comparan los resultados del **paquete [cvxpy](https://github.com/cvxgrp/cvxpy)** con los obtenidos en la implementación hecha por el prof en [algoritmos/Python](algoritmos/Python), en específico [algoritmos/Python/algorithms_for_ceco.py](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/algoritmos/Python/algorithms_for_ceco.py) para problemas tipo CECO (*Constrained Equality Convex Optimization*)

In [1]:
import os

In [2]:
cur_directory = os.getcwd()

In [3]:
dir_alg_python = '/algoritmos/Python'

In [4]:
os.chdir(cur_directory + dir_alg_python)

In [5]:
import math

import numpy as np

import algorithms_for_ceco
from line_search import line_search_for_residual_by_backtracking
from utils import compute_error


# Primer ejemplo

$$\displaystyle \min_{x \in \mathbb{R}^2} \quad x_1^2 + 2x_1x_2 + x_2^2-2x_2$$

$$\text{sujeto a: } x_1 = 0$$

In [6]:
fo = lambda x: x[0]**2 + 2*x[0]*x[1]+x[1]**2-2*x[1]

In [7]:
A = np.array([1,0],dtype=float)

In [8]:
b = np.array([0])

In [9]:
x_ast=np.array([0,1], dtype=float)

## Punto inicial $x^{(0)}$ factible ($Ax^{(0)}=b$)

In [10]:
x_0 = np.array([0,-2],dtype=float)

In [11]:
tol=1e-8
tol_backtracking=1e-14
maxiter=50
p_ast=fo(x_ast)
[x,total_of_iterations,
 Err_plot,x_plot]=algorithms_for_ceco.Newtons_method_feasible_init_point(fo,A, x_0,tol, 
                                                                         tol_backtracking, x_ast, p_ast, maxiter)

I	Normgf 	Newton Decrement	Error x_ast	Error p_ast	line search	CondHf
0	7.21e+00	1.80e+01	3.00e+00	9.00e+00	---		9.00e+03
1	7.21e+00	1.15e-05	2.40e-03	5.76e-06	1.00e+00	9.00e+03
2	7.21e+00	7.55e-15	6.68e-08	4.44e-15	1.00e+00	9.00e+03
Error of x with respect to x_ast: 6.68e-08
Approximate solution: [4.33680869e-19 9.99999933e-01]


In [12]:
x

array([4.33680869e-19, 9.99999933e-01])

In [13]:
total_of_iterations

3

In [14]:
x_plot.shape

(2, 3)

In [15]:
x_plot

array([[ 0.00000000e+00,  0.00000000e+00,  4.33680869e-19],
       [-2.00000000e+00,  1.00239973e+00,  9.99999933e-01]])

In [16]:
compute_error(x_ast,x)

6.682210107467057e-08

## En el siguiente ejemplo si usamos el algoritmo anterior para punto inicial no factible ($Ax^{(0)} \neq b$) obsérvese que no funciona

In [17]:
x_0 = np.array([1,-2],dtype=float)

In [18]:
tol=1e-8
tol_backtracking=1e-14
maxiter=50
p_ast=fo(x_ast)
[x,total_of_iterations,
 Err_plot,x_plot]=algorithms_for_ceco.Newtons_method_feasible_init_point(fo,A, x_0,tol, 
                                                                         tol_backtracking, x_ast, p_ast, maxiter)

I	Normgf 	Newton Decrement	Error x_ast	Error p_ast	line search	CondHf
0	4.47e+00	8.01e+00	3.16e+00	6.00e+00	---		2.03e+07
1	4.47e+00	5.12e-06	1.41e+00	2.00e+00	1.00e+00	2.03e+07
2	4.47e+00	9.86e-16	1.41e+00	2.00e+00	1.00e+00	2.03e+07
Error of x with respect to x_ast: 1.41e+00
Approximate solution: [ 1.00000000e+00 -3.34455919e-08]


In [19]:
compute_error(x_ast,x)

1.4142135860226999

**Error relativo muy alto**

# Segundo ejemplo

$$\displaystyle \min_{x \in \mathbb{R}^2} \quad \frac{1}{2}(x_1^2 + x_2^2)$$

$$\text{sujeto a: } 2x_1 -x_2 = 5$$

In [20]:
fo = lambda x: 1/2*(x[0]**2  + x[1]**2)

In [21]:
A = np.array([2,-1],dtype=float)

In [22]:
b=5

In [23]:
x_ast = np.array([2,-1],dtype=float)

## Punto inicial $x^{(0)}$ factible ($Ax^{(0)}=b$)

In [24]:
x_0 = np.array([0,-5],dtype=float)

In [25]:
tol=1e-8
tol_backtracking=1e-14
maxiter=50
p_ast=fo(x_ast)
[x,total_of_iterations,
 Err_plot,x_plot]=algorithms_for_ceco.Newtons_method_feasible_init_point(fo,A, x_0,tol, 
                                                                         tol_backtracking, x_ast, p_ast, maxiter)

I	Normgf 	Newton Decrement	Error x_ast	Error p_ast	line search	CondHf
0	5.00e+00	2.00e+01	2.00e+00	4.00e+00	---		1.00e+00
1	5.00e+00	3.94e-06	8.88e-04	7.89e-07	1.00e+00	1.00e+00
2	5.00e+00	3.56e-12	8.63e-07	7.45e-13	1.00e+00	1.00e+00
Error of x with respect to x_ast: 8.63e-07
Approximate solution: [ 1.99999914 -1.00000173]


In [26]:
x

array([ 1.99999914, -1.00000173])

In [27]:
x_ast

array([ 2., -1.])

In [28]:
compute_error(x_ast,x)

8.631750624932266e-07

# Tercer ejemplo

$$\displaystyle \min_{x \in \mathbb{R}^2} \quad x_1^2 + x_2^2$$

$$\text{sujeto a :} x_1+x_2 = 1$$

In [29]:
fo = lambda x: x[0]**2+x[1]**2

In [30]:
x_ast = np.array([.5,.5],dtype=float)

In [31]:
A = np.array([1,1],dtype=float)

In [32]:
b=1

## Punto inicial $x^{(0)}$ factible ($Ax^{(0)}=b$)

In [33]:
x_0 = np.array([2,-1],dtype=float)

In [34]:
tol=1e-8
tol_backtracking=1e-14
maxiter=50
p_ast=fo(x_ast)
[x,total_of_iterations,
 Err_plot,x_plot]=algorithms_for_ceco.Newtons_method_feasible_init_point(fo,A, x_0,tol, 
                                                                         tol_backtracking, x_ast, p_ast, maxiter)

I	Normgf 	Newton Decrement	Error x_ast	Error p_ast	line search	CondHf
0	4.47e+00	9.00e+00	3.00e+00	9.00e+00	---		1.00e+00
1	4.47e+00	7.11e-08	2.67e-04	7.11e-08	1.00e+00	1.00e+00
2	4.47e+00	1.81e-16	2.46e-12	1.11e-16	1.00e+00	1.00e+00
Error of x with respect to x_ast: 2.46e-12
Approximate solution: [0.5 0.5]


In [35]:
compute_error(x_ast,x)

2.4647506264441902e-12

# Cuarto ejemplo

$$\displaystyle \min_{x \in \mathbb{R}^2} \quad e^{x_1+3x_2-0.1} + e^{x_1 -3x_2-0.1} + e^{-x_1-0.1}$$

$$\text{sujeto a:} x_1 + 3x_2 = 0$$

In [44]:
fo = lambda x: math.exp(x[0] + 3*x[1]-0.1) + math.exp(x[0]  -3*x[1]-0.1) + math.exp(-x[0]-0.1)

In [45]:
x_ast = np.array([-0.23104907880100917,0.0770163596518852],dtype=float)

In [46]:
A = np.array([1,3],dtype=float)

In [47]:
b=0

## Punto inicial $x^{(0)}$ factible ($Ax^{(0)}=b$)

In [48]:
x_0 = np.array([0,0],dtype=float)

In [49]:
tol=1e-8
tol_backtracking=1e-14
maxiter=50
p_ast=fo(x_ast)
[x,total_of_iterations,
 Err_plot,x_plot]=algorithms_for_ceco.Newtons_method_feasible_init_point(fo,A, x_0,tol, 
                                                                         tol_backtracking, x_ast, p_ast, maxiter)

I	Normgf 	Newton Decrement	Error x_ast	Error p_ast	line search	CondHf
0	9.05e-01	1.81e-01	1.00e+00	3.81e-02	---		6.00e+00
1	9.05e-01	3.30e-03	1.34e-01	6.37e-04	1.00e+00	6.00e+00
2	9.05e-01	8.46e-07	2.15e-03	1.62e-07	1.00e+00	6.00e+00
3	9.05e-01	7.41e-14	6.52e-07	1.98e-11	1.00e+00	6.00e+00
Error of x with respect to x_ast: 6.52e-07
Approximate solution: [-0.23104893  0.07701631]


In [50]:
compute_error(x_ast, x)

6.524588707146708e-07

# Quinto ejemplo: con más restricciones de igualdad

Resolver: 
$$\displaystyle \min_{x \in \mathbb{R}^3} \quad ||x||_2^2$$

$$\text{sujeto a: }\begin{array}{l}
\begin{array}{c}
x_1 + x_2 + x_3 = 1 \\
x_1 + x_2 + 2x_3 = 3
\end{array}
\end{array}
$$

In [51]:
fo = lambda x: x.dot(x)

In [52]:
x_ast = np.array([-0.5,-0.5,2. ], dtype=float)

In [53]:
A = np.array([[1, 1, 1],
              [1, 1, 2]],dtype=float)

In [54]:
b = np.array([1,3])

## Punto inicial $x^{(0)}$ factible ($Ax^{(0)}=b$)

In [55]:
x_0 = np.array([3,-4,2], dtype=float)
#x_0 = np.array([2,-3,2], dtype=float) #este es otro punto factible

In [56]:
tol=1e-8
tol_backtracking=1e-14
p_ast=fo(x_ast)


In [57]:
[x,total_of_iterations,
 Err_plot,x_plot]=algorithms_for_ceco.Newtons_method_feasible_init_point(fo,A, x_0,tol, 
                                                                         tol_backtracking, x_ast, p_ast, maxiter)

I	Normgf 	Newton Decrement	Error x_ast	Error p_ast	line search	CondHf
0	1.08e+01	4.90e+01	2.33e+00	5.44e+00	---		1.00e+00
1	1.08e+01	3.14e-05	1.87e-03	3.48e-06	1.00e+00	1.00e+00
2	1.08e+01	5.83e-13	2.52e-07	6.45e-14	1.00e+00	1.00e+00
Error of x with respect to x_ast: 2.52e-07
Approximate solution: [-0.49999962 -0.50000038  2.        ]


In [58]:
compute_error(x_ast,x)

2.518419818612709e-07

# Comparación del quinto ejemplo con [cvxpy](https://github.com/cvxgrp/cvxpy)

In [62]:
import cvxpy as cp

In [63]:
x = cp.Variable(3)
A = np.array([[1, 1, 1],
              [1, 1, 2]])
b = np.array([1,3])

In [64]:
obj = cp.Minimize(cp.norm(x,2))

constraints = [A@x == b]

In [65]:
prob = cp.Problem(obj, constraints)

In [66]:
prob.solve()

2.1213203277662585

In [67]:
print("optimal value", prob.value)


optimal value 2.1213203277662585


In [68]:
print("optimal var", x.value)

optimal var [-0.5 -0.5  2. ]


In [69]:
A@x.value

array([1., 3.])

In [70]:
np.linalg.norm(x.value)

2.121320343558957

## Otros puntos iniciales

In [71]:
vec = [1,-2,2]
A@vec

array([1, 3])

In [72]:
np.linalg.norm(vec)

3.0

In [73]:
vec = [2,-3,2]
A@vec

array([1, 3])

In [74]:
np.linalg.norm(vec)

4.123105625617661

In [75]:
vec = [3,-4,2]
A@vec

array([1, 3])

In [76]:
np.linalg.norm(vec)

5.385164807134504

---

# Definiciones utilizadas en el curso (algunas de ellas...)

A partir de esta sección consideramos problemas de optimización de la forma:

$$ \min f_o(x)$$

$$\text{sujeto a: } f_i(x) \leq 0 \quad \forall i=1,\dots,m$$

$$h_i(x) = 0 \quad \forall i=1,\dots,p$$

donde: $f_o$, $f_i \forall i=1,\dots,m$ y $h_i \forall i=1,\dots,p$ son funciones diferenciables. Este problema lo nombramos **problema primal de optimización**.

**Obs:** obsérvese que **no** estamos asumiendo que tenemos un problema convexo.

# La función Lagrangiana $\mathcal{L}$

Definimos la **función Lagrangiana** asociada al problema de optimización (primal) como: $\mathcal{L}: \mathbb{R}^n \times \mathbb{R}^m \times \mathbb{R}^p \rightarrow \mathbb{R}$ con:

$$\mathcal{L}(x, \lambda , \nu) = f_o(x) + \displaystyle \sum_{i=1}^m \lambda_i f_i(x) + \sum_{i=1}^p \nu_i h_i(x)$$

y $\text{dom} \mathcal{L} = \mathcal{D} \times \mathbb{R}^m \times \mathbb{R}^p$ donde: $\mathcal{D}$ es el dominio del problema de optimización. Ver el apéndice de la nota [4.1.Optimizacion_numerica_y_machine_learning](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.1.Optimizacion_numerica_y_machine_learning.ipynb) para definición del dominio de un problema de optimización.

**Comentarios:**

* $\lambda _i$ se le nombra **multiplicador de Lagrange** asociado con la $i$-ésima restricción de desigualdad $f_i(x) \leq 0$. 

* $\nu_i$ se le nombra **multiplicador de Lagrange** asociado con la $i$-ésima restricción de igualdad $h_i(x)=0$.

* Los vectores $\lambda = (\lambda_i)_{i=1}^m$ y  $\nu = (\nu_i)_{i=1}^p \in \mathbb{R}^p$ se les nombran **variables duales** o **vectores de multiplicadores de Lagrange** asociados con el problema de optimización. El vector $x \in \mathcal{D}$ se le nombra **variable primal**.


**Consideramos en lo que continúa la restricción $\lambda_i \geq 0 \forall i=1,\dots, m$**.


Podemos interpretar a la función Lagrangiana como una **subestimación relajada** del problema primal de optimización. Tal problema puede reescribirse en uno **sin restricciones** de la forma:


$$\min f_o(x) + \displaystyle \sum_{i=1}^m I_{-}(f_i(x)) + \sum_{i=1}^p I_o(h_i(x))$$

donde:

$$I_{-}(u) = \begin{cases}
0 & \text{si } u \leq 0,\\
\infty & \text{si } u > 0\\
\end{cases}
\quad 
I_o(u) = \begin{cases}
0 & \text{si } u=0,\\
\infty & \text{si } u \neq 0\\
\end{cases}
$$

**Obs:** para ver la equivalencia entre este problema **sin restricciones** y el primal piénsese en que a la función $f_o$ se le está añadiendo una **regularización** en donde se penaliza fuertemente a los vectores $x$ que **no** cumplen con $f_i(x) \leq 0$ o $h_i(x) = 0$. Ver [4.3.Minimos_cuadrados_R](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/IV.optimizacion_convexa_y_machine_learning/4.3.Minimos_cuadrados_R.ipynb) para regularización en un problema de mínimos cuadrados y la relación de la regularización con problemas *bi-criterion* de optimización.

Como $\lambda_i \geq 0$ se tiene: 


$$\lambda_i f_i(x) \leq I_{-}(f_i(x)) \quad \forall i=1,\dots,m$$ 

$$\nu_ih_i(x) \leq I_o(h_i(x)) \quad \forall i=1,\dots,p.$$

La Lagrangiana $\mathcal{L}(x, \nu) = f_o(x) + \displaystyle \sum_{i=1}^m \nu_i h_i(x)$ entonces cumple $\mathcal{L}(x, \nu) \leq f_o(x) + \displaystyle \sum_{i=1}^p I_o(h_i(x))$ $\forall (x, \nu ) \in \text{dom}\mathcal{L}$. Esto es, $\mathcal{L}(x,\nu)$ subestima a $f_o(x) + \displaystyle \sum_{i=1}^m \lambda_i f_i(x) +  \sum_{i=1}^m \nu_i h_i(x)$ .

# Función dual de Lagrange

La función dual de Lagrange se define como $g: \mathbb{R}^m \times \mathbb{R}^p \rightarrow \mathbb{R}$ con:

$$g(\lambda, \nu) = \displaystyle \inf_{x \in \mathcal{D}} \mathcal{L}(x,\lambda, \nu)$$

**Comentarios:**

* Recuérdese que $\inf$ es el máximo de las cotas inferiores, ver [1.3.Condicion_de_un_problema_y_estabilidad_de_un_algoritmo](https://github.com/ITAM-DS/analisis-numerico-computo-cientifico/blob/master/temas/I.computo_cientifico/1.3.Condicion_de_un_problema_y_estabilidad_de_un_algoritmo.ipynb) para definición de supremo e ínfimo. En esta definción se está tomando el ínfimo de la Lagrangiana para cada punto $x \in \mathcal{D}$ (ínfimo puntual).

* Por la definición de la función dual de Lagrange y el punto anterior, se tiene: $g(\lambda, \nu)  \leq \mathcal{L}(\tilde{x},\lambda, \nu)$ para $\tilde{x}$ factible.

* La función dual de Lagrange para cada par $(\lambda, \nu)$ con $\lambda \geq 0$ da una **subestimación de $p^*$, el valor óptimo del problema primal**. Esto es: $g(\lambda, \nu) \leq p^*$ para $\lambda \geq 0, \nu \in \mathbb{R}^p$ y $p^*$ valor óptimo del problema de optimización primal. 

* Si la Lagrangiana no es acotada por debajo entonces la función dual de Lagrange se asigna $-\infty$.

* Se prueba que la función dual de Lagrange **siempre es una función cóncava**.



## Problema dual de Lagrange

Asociado a **todo** problema de optimización existe su **problema dual** el cual se define como:

$$\displaystyle \max g(\lambda,\nu)$$

$$\text{sujeto a: } \lambda \geq 0.$$

**Comentarios:**


* Si $(\lambda, \nu)$ satisface $\lambda \geq 0$ y $g(\lambda, \nu)$ no es $-\infty$ la pareja $(\lambda, \nu)$ se nombra **dual factible**.

* El **valor óptimo del problema dual** se denota con $d^*$. **El valor óptimo se alcanza** si existen $(\lambda^*, \nu^*)$ tales que $d^* = g(\lambda^*, \nu^*)$. 

* La pareja $(\lambda^*, \nu^*)$ se nombra **óptimo dual** o **pareja óptima de multiplicadores de Lagrange** si satisface:

    * $\lambda^* \geq 0$, 
    * $g(\lambda^*, \nu^*)$ no es $-\infty$ y 
    * $(\lambda^*, \nu^*)$ resuelve el problema dual. 


* Para distinguir a los óptimos duales de los puntos óptimos $x^*$ del problema primal, a éste últimos se les nombra **óptimos primales**.

* El problema dual tiene como interpretación **la mejor cota inferior de la función Lagrangiana como función de $x$ que se obtiene con la función dual de Lagrange**.

* El problema dual de Lagrange **siempre es un problema de optimización convexo aún si el problema primal no es convexo.**

* A la diferencia $p^* - d^*$ se le nombra ***optimal duality gap*.** Si es positiva se tiene un **problema de optimización en el que se cumple dualidad débil**. Si es igual a cero se tiene un **problema de optimización en el que se cumple dualidad fuerte**.

## Dualidad fuerte y holgura complementaria

En un problema de optimización en el que los valores óptimos primal y dual se alcanzan y en el que se cumple dualidad fuerte se tiene la propiedad que el **punto óptimo primal minimiza $\mathcal{L}(x,\lambda^*, \nu^*)$** pues:

$$f_o(x^*)  = p^* = d^* = g(\lambda^*, \nu^*) = \inf_{x \in \mathcal{D}} \mathcal{L}(x, \lambda^*, \nu^*)$$

**Comentario:** lo anterior prueba que $\lambda^*_i f_i(x^*)=0 \quad  \forall i=1, \dots, m$. Tal ecuación se conoce con el nombre de **holgura complementaria**. La holgura complementaria indica que si el $i$-ésimo multiplicador de Lagrange es positivo entonces la $i$-ésima función de restricción de desigualdad $f_i$ es activa:

$$\lambda^*_i > 0 \implies f_i(x^*) = 0$$

$$f_i(x^*) < 0 \implies \lambda^*_i = 0.$$

## Condiciones KKT para problemas de optimización (convexos o no)

Sean $x^*, (\lambda^*, \nu^*)$ un par primal-dual de puntos óptimos en los que la *optimal duality gap* es cero o bien: $d^* = p^*$. Entonces $x^*$ minimiza $\mathcal{L}(x, \lambda^*, \nu^*)$ y por factibilidad y holgura complementaria se cumple:

$$\nabla_x \mathcal{L}(x^*, \lambda^*, \nu^*) = \nabla f_o(x^*) + \displaystyle \sum_{i=1}^m \lambda^*_i f_i(x^*)  + \sum_{i=1}^p \nu_i^* \nabla h_i(x^*) = 0$$

$$f_i(x^*) \leq 0 \quad \forall i=1,\dots,m$$

$$h_i(x^*)=0 \quad \forall i = 1,\dots, p.$$

$$\lambda_i^* \geq 0 \quad \forall i =1,\dots,m$$

$$\lambda^*_if_i(x^*) = 0 \quad \forall i=1,\dots,m.$$

Las $5$ condiciones anteriores se nombran **condiciones KKT de optimalidad** para el problema de optimización del inicio. 

**Comentarios:**

* Las condiciones KKT de optimalidad se satisfacen para **cualquier problema de optimización con función objetivo  y funciones de restricción diferenciables evaluadas en un par primal-dual de puntos óptimos en el que se cumple la dualidad fuerte**. 

* La notación $\nabla_x$ se refiere a la derivada de $\mathcal{L}$ respecto a la variable $x$.

*  Son condiciones necesarias.

* La ecuación $h_i(x^*)=0 \forall i=1,\dots,p$  y $f_i(x^*) \leq 0 \forall i=1,\dots,m$ se nombran **ecuaciones de factibilidad primal** y $\nabla_x \mathcal{L}(x^*,\lambda^*, \nu^*)=0$ se nombra **ecuación de factibilidad dual**.


* La dualidad fuerte y que el óptimo del problema dual se alcance puede establecerse a partir de condiciones teóricas sobre el problema de optimización primal. Un ejemplo son las de Slater, ver [Slater's condition](https://en.wikipedia.org/wiki/Slater%27s_condition). Para condiciones más generales aplicables a problemas de optimización primal no necesariamente convexo existen las **condiciones de regularidad**, ver [regurlarity conditions or constraint qualifications](https://en.wikipedia.org/wiki/Karush%E2%80%93Kuhn%E2%80%93Tucker_conditions#Regularity_conditions_(or_constraint_qualifications)).

## Condiciones KKT para problemas de optimización convexos

Si el problema **primal** es convexo, esto es, $f_o, f_1, \dots, f_m$ son funciones convexas y $h_i$ es afín $\forall i,\dots, p$ entonces las condiciones KKT son **suficientes** para puntos primal-dual óptimos:

Sean $\tilde{x}, \tilde{\lambda}, \tilde{\nu}$ puntos que satisfacen las condiciones KKT de optimalidad:

$$\nabla_x \mathcal{L}(\tilde{x}, \tilde{\lambda}, \tilde{\nu}) = \nabla f_o(\tilde{x})  +  \displaystyle \sum_{i=1}^m \tilde{\lambda}_i f_i(\tilde{x}) + \sum_{i=1}^p \tilde{\nu}_i \nabla h_i(\tilde{x}) = 0$$


$$f_i(\tilde{x}) \leq 0 \quad \forall i=1,\dots,m$$

$$h_i(\tilde{x})=0 \quad \forall i = 1,\dots, p.$$

$$\tilde{\lambda_i} \geq 0 \quad \forall i=1,\dots,m$$

$$\tilde{\lambda}_if_i(\tilde{x}) = 0 \quad \forall i=1,\dots,m.$$

**entonces el par $(\tilde{x},\tilde{\nu})$ son puntos primal-dual óptimos y en el problema de optimización se cumple la dualidad fuerte.**

**Comentario:** en algunos casos especiales es posible resolver las condiciones KKT de optimalidad de manera analítica y en general muchos algoritmos de optimización resultan o pueden interpretarse como métodos que las resuelven.

**Referencias:**

* S. P. Boyd, L. Vandenberghe, Convex Optimization, Cambridge University Press, 2009.